website/docs/docs/build/udfs.md

```diff
@@ -0,0 +1,208 @@
+        arguments: # optional
+            - name: a_string # required if arguments is specified
+              data_type: string # required if arguments is specified
+              description: The string that I want to check if it's representing a positive integer (like "10") 
+        returns: # required
+            data_type: boolean # required
+    ```
+    </File>
+
+    <!--other types not yet supported
+    <Expandable alt_header="Supported UDF types">
+
+    You can use these values for the `type` property when you define a function in a YAML file.
+
+    - `scalar` - Returns a single value per row
+    - `aggregate` - Returns a single value per group, aggregating several rows
+    - `table functions` - Returns a table result
+    <br></br>
+    For example:
+
+    ```yml
+    functions:
+	  - name: string 
+	    description: string
+	    type: scalar # default value
+    ```
+
+    If not explicitly specified, the `type` property defaults to `scalar`.
+
+    </Expandable>
+    -->
+    
+    The rendered `CREATE` UDF statement depends on which adapter you’re using. For example:
+
+    <Tabs>
+
+    <TabItem value="Snowflake">
+    ```sql
+    CREATE OR REPLACE FUNCTION my_schema.is_positive_int(a_string STRING)
+    RETURNS BOOLEAN
+    AS (
+    REGEXP_CONTAINS(a_string, r'^[0-9]+$')
+    );
+    ```
+    </TabItem>
+
+    <TabItem value="Redshift">
+    ```sql
+    CREATE OR REPLACE FUNCTION repeat_n(s VARCHAR, n INTEGER)
+    RETURNS VARCHAR
+    VOLATILE
+    AS $$ SELECT repeat(s, n); $$
+    LANGUAGE SQL;
+    ```
+    </TabItem>
+    </Tabs>
+
+3. Reference the UDF in a model using the `{{ function(...) }}` macro. For example:
+
+    <File name="models/my_model.sql">
+
+    ```sql
+    select
+
+    maybe_positive_int_column,
+        {{ function('is_positive_int') }}(maybe_positive_int_column)
+
+    from {{ ref('a_model_i_like') }}
+    ```
+    </File>
+
+4. Run `dbt compile`. In the following example, the `{{ function('is_positive_int') }}` is replaced by the UDF name `udf_db.udf_schema.is_positive_int`.
+
+    <File name="models/my_model.sql">
+
+    ```sql
+    select
+
+    maybe_positive_int_column,
+	udf_db.udf_schema.is_positive_int(maybe_positive_int_column) as is_positive
+
+    from analytics.<dbt_schema>.a_model_i_like
+    ```
+    </File>
+
+    In your DAG, there should be a dependency between `is_positive_int` → `my_model` and a UDF node is created from the SQL and YAML definitions.
+
+After defining a UDF, you can update the SQL file that contains its function body (`is_positive_int.sql` in this example). When you rebuild, your changes will be applied everywhere the UDF is referenced. 
+
+## Using UDFs in unit tests
+
+You can use [unit tests](/docs/build/unit-tests) to validate models that reference UDFs. Before running unit tests, make sure the function exists in your warehouse. To ensure that the function exists for a unit test, run:
+
+```bash
+dbt build --select "+my_model_to_test" --empty
```

Comment:
@QMalcolm does `--empty` work with functions? that doesn't make sense to me

Reply:
@graciegoheen Empty doesn't do _anything_ to functions. However, most unit tests also reference _other models_. For unit tests that depend on other models, the upstream models need to exists (so we can get some information about them). Because of that, we recommend `dbt run --select "+my_model_to_test" --empty` being run first so that the upsteam models exist (but aren't costly in insuring their existence). Doing the `build` in the documentation will also build the upstream models, and thus we should still recommend `--empty`